neqo-transport/src/fc.rs

```diff
@@ -330,17 +378,51 @@ impl ReceiverFlowControl<()> {
         builder: &mut packet::Builder<B>,
         tokens: &mut recovery::Tokens,
         stats: &mut FrameStats,
+        now: Instant,
+        rtt: Duration,
     ) {
         if !self.frame_needed() {
             return;
         }
+
+        self.auto_tune(now, rtt);
+
         let max_allowed = self.next_limit();
         if builder.write_varint_frame(&[FrameType::MaxData.into(), max_allowed]) {
             stats.max_data += 1;
             tokens.push(recovery::Token::Stream(StreamRecoveryToken::MaxData(
                 max_allowed,
             )));
             self.frame_sent(max_allowed);
+            self.last_update = Some(now);
+        }
+    }
+
+    /// Auto-tune [`ReceiverFlowControl::max_active`], i.e. the connection flow
+    /// control window.
+    ///
+    /// If the sending rate (`window_bytes_used`) exceeds the rate allowed by
+    /// the maximum flow control window and the current rtt
+    /// (`window_bytes_expected`), try to increase the maximum flow control
+    /// window ([`ReceiverFlowControl::max_active`]).
+    fn auto_tune(&mut self, now: Instant, rtt: Duration) {
+        let prev_max_active = self.max_active;
+        let increase = self.auto_tune_impl(now, rtt, MAX_CONN_RECV_WINDOW_SIZE);
+
+        if increase > 0 {
+            qdebug!(
+                "Increasing max connection receive window by {} B, \
+                previous max_active: {} MiB, \
+                new max_active: {} MiB, \
+                last update: {:?}, \
+                rtt: {rtt:?}",
+                increase,
+                prev_max_active / 1024 / 1024,
+                self.max_active / 1024 / 1024,
+                now - self
+                    .last_update
+                    .expect("last_update must be set by auto_tune_impl"),
+            );
```

Comment:
This only differs between connection and stream auto-tuning in the template string "Increasing max connection receive window", right? What do you think of passing `"connection"` and `"stream"` to `auto_tune_inner` and then add it to the template string:

``` rust
            qdebug!(
                "Increasing max {} receive window by {} B, \